File: Structural/Bridge/java/Bridge.java
package Structural.Bridge;

/**
 * Defines the abstraction in the Bridge Pattern.
 * Works independently of the implementation.
 */
abstract class Abstraction {
	protected Implementor implementor;

	/**
	 * Constructs the abstraction with a given implementor.
	 * @param implementor The concrete implementor to bridge with.
	 */
	public Abstraction(Implementor implementor) {
		this.implementor = implementor;
	}

	/**
	 * Performs an operation, to be defined by subclasses.
	 */
	public abstract void operation();
}

/**
 * Refined abstraction that extends the base abstraction.
 */
class RefinedAbstraction extends Abstraction {

	/**
	 * Constructs the refined abstraction with a given implementor.
	 * @param implementor The concrete implementor to bridge with.
	 */
	public RefinedAbstraction(Implementor implementor) {
		super(implementor);
	}

	/**
	 * Performs the operation using the implementor.
	 */
	@Override
	public void operation() {
		System.out.println("RefinedAbstraction: Delegating operation to implementor.");
		implementor.implementation();
	}
}

/**
 * Defines the interface for the implementor in the Bridge Pattern.
 */
interface Implementor {
	/**
	 * Defines the operation implemented by concrete implementors.
	 */
	void implementation();
}

/**
 * Concrete implementor that provides a specific implementation.
 */
class ConcreteImplementorA implements Implementor {

	/**
	 * Provides a concrete implementation of the operation.
	 */
	@Override
	public void implementation() {
		System.out.println("ConcreteImplementorA: Performing implementation.");
	}
}

/**
 * Another concrete implementor that provides a different implementation.
 */
class ConcreteImplementorB implements Implementor {

	/**
	 * Provides another concrete implementation of the operation.
	 */
	@Override
	public void implementation() {
		System.out.println("ConcreteImplementorB: Performing implementation.");
	}
}

File: Structural/Bridge/python/bridge.py
import logging
from abc import ABC, abstractmethod

# Configure the logging module
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class Implementor(ABC):
	"""
	Defines the interface for the implementation in the Bridge Pattern.
	"""

	@abstractmethod
	def implementation(self) -> None:
		"""
		Defines the operation implemented by concrete implementors.
		"""
		pass

class ConcreteImplementorA(Implementor):
	"""
	Provides a specific implementation of the Implementor interface.
	"""

	def implementation(self) -> None:
		"""
		Provides a concrete implementation of the operation.
		"""
		logger.info("ConcreteImplementorA: Performing implementation.")

class ConcreteImplementorB(Implementor):
	"""
	Provides another specific implementation of the Implementor interface.
	"""

	def implementation(self) -> None:
		"""
		Provides a concrete implementation of the operation.
		"""
		logger.info("ConcreteImplementorB: Performing implementation.")

class Abstraction:
	"""
	Defines the abstraction in the Bridge Pattern.
	Works independently of the implementation.
	"""

	def __init__(self, implementor: Implementor) -> None:
		"""
		Initializes the abstraction with a given implementor.

		:param implementor: The concrete implementor to bridge with.
		"""
		self._implementor = implementor

	def operation(self) -> None:
		"""
		Performs an operation using the implementor.
		"""
		logger.info("Abstraction: Delegating operation to implementor.")
		self._implementor.implementation()

class RefinedAbstraction(Abstraction):
	"""
	Extends the base abstraction to add more behavior.
	"""

	def operation(self) -> None:
		"""
		Adds additional behavior before or after delegating to the implementor.
		"""
		logger.info("RefinedAbstraction: Additional behavior before delegation.")
		super().operation()
		logger.info("RefinedAbstraction: Additional behavior after delegation.")
